### PR TITLE
Use next slot cache for payload attribute

### PR DESCRIPTION
**Pull request by terencechain**
This PR changes
- Use next slot cache for `getPayloadAttribute` to avoid duplicated `ProcessSlotsIfPossible` if necessary. This becomes relevant for epoch boundary blocks

### CODE REVIEW

File: beacon-chain/blockchain/execution_engine.go [hunk 1]
```diff
@@ -265,7 +265,8 @@ func (s *Service) getPayloadAttribute(ctx context.Context, st state.BeaconState,
 
 	// Get previous randao.
 	st = st.Copy()
-	st, err := transition.ProcessSlotsIfPossible(ctx, st, slot)
+	headRoot := s.headRoot()
+	st, err := transition.ProcessSlotsUsingNextSlotCache(ctx, st, headRoot[:], slot)
```
Comment (potuz) [inline beacon-chain/blockchain/execution_engine.go:269]: There is a functional difference between `ProcessSlotsIfPossible` and using NSC: the former does not error out if the requested slot is higher than the current slot. I've seen this fail on tests and I wonder if this won't fail on runtime.

Comment (terencechain) [reply beacon-chain/blockchain/execution_engine.go:269]: Right. The easiest way to solve this is to add process slots only if it's higher within `ProcessSlotsUsingNextSlotCache`. It seems no harm to do that but curious what you think

File: beacon-chain/core/transition/trailing_slot_state_cache.go [hunk 2]
```diff
@@ -14,7 +14,7 @@ import (
 )
 
 type nextSlotCache struct {
-	sync.Mutex
+	sync.RWMutex
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/core/transition/trailing_slot_state_cache.go:17]: No dead lock but curious why we don't use `RWMutex` to begin with? It seems inevitable we move to that direction

**Review by potuz — DISMISSED** [review]
Approving with a comment, should we error out if `slot < st.Slot()`?

File: beacon-chain/blockchain/execution_engine.go [hunk 3]
```diff
@@ -261,10 +261,13 @@ func (s *Service) getPayloadAttribute(ctx context.Context, st state.BeaconState,
 
 	// Get previous randao.
 	st = st.Copy()
-	st, err := transition.ProcessSlotsIfPossible(ctx, st, slot)
-	if err != nil {
-		log.WithError(err).Error("Could not process slots to get payload attribute")
-		return false, emptyAttri, 0
+	if slot > st.Slot() {
+		var err error
+		st, err = transition.ProcessSlotsUsingNextSlotCache(ctx, st, headRoot, slot)
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/execution_engine.go:266]: This doesnt seem right. The method signature for ProcessSlotsUsingNextSlotCache requires the parent root in the 3rd argument, not the head root? If that is the case, why not read the parent root of `st`?

https://github.com/prysmaticlabs/prysm/blob/fa7a2bd44589a4c9861deff3d5157fe0f03bb0d9/beacon-chain/core/transition/transition.go#L142-L146

Comment (prestonvanloon) [reply beacon-chain/blockchain/execution_engine.go:266]: ah nvm -- it calls for the parent state too

**Review by prestonvanloon — APPROVED** [review]
Code seems good. I didn't run it locally

**Pull request merged**